Pull request 658

robcamer opened the pull request:
Azure Automation Account module for terratest with the following:

(i) terraform example in examples/azure folder
(ii) automationaccount.go module in modules/azure folder
(iii) unit test in test/azure folder
(iv) integration test with example in test/azure folder

NOTE: This PR updates the workflow  `ci.yml` to run tests using the  `azure` +build tag.  To validate this PR you must run the ci.yml in this branch as the ci.yml in master will break due to the automation account tests.

This PR adds a new ci workflow `ci-automation.yml` for the Automation Account tests that uses an `azureautomation` +build tag.  The default `ci.yml` is updated to run the go linter tests based on the `azure` build tag:
LINTER:
./bin/golangci-lint run ./test/azure/ --build-tags=azure --timeout 5m0s
TESTS:
 go test -tags=azure  -v -timeout 90m

The new `ci-automation.yml` has a build tag of `azureuatomation`
LINTER:
./bin/golangci-lint run ./test/azure/ --build-tags=azureautomation --timeout 5m0s
TEST:
go test -tags=azureautomation -v -timeout 20m


This example requires a service principal and test certificate in order to run successfully. This is documented in the readme.md for the example:
"_In this case for the test example, the test self-signed certificate is base64 encoded and then set on the `TF_VAR_automation_run_as_certificate_base64` environment variable.  As an alternative, you could read the certificate from Key Vault as described above, which would be the recommended procedure in a production environment._"

Fixes #658

polatengin committed:
Update ci.yml

rguthriemsft committed:
Merge remote-tracking branch 'upstream/master'

HadwaAbdelhalem committed:
Merge remote-tracking branch 'upstream/master' into master

rguthriemsft committed:
Merge branch 'master' of github.com:aztfmod/terratest

hattan committed:
Adding action to set up Terraform 14.9

hattan committed:
Update CI to pin to terraform 15.1

hattan committed:
Merge remote-tracking branch 'upstream/master'

hattan commented:
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/aztfmod/terratest/actions/runs/804759161

hattan committed:
Merge remote-tracking branch 'upstream/master'

HadwaAbdelhalem committed:
Merge remote-tracking branch 'upstream/master' into master

HadwaAbdelhalem committed:
Merge remote-tracking branch 'upstream/master' into master

rguthriemsft committed:
Merge remote-tracking branch 'upstream/master'

robcamer committed:
rebase on master

hattan commented:
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/aztfmod/terratest/actions/runs/837150407

yorinasub17 committed:
Fix build (#898)

* Bump to go 1.16

* Limit build parallelism

* Add go111module setting

* Use go 1.16.3

* Use working directory in go src dir

* Make sure all jobs have the same working_directory

* Call go mod tidy

* bump hardcoded go version in test fixtures

yorinasub17 committed:
Fix bug in terratest-log-parser that mixes result lines (#899)

Commit:
Replace testing.T with testing.TestingT (#904)

Commit:
Fix typos in the comments

cbascom committed:
Replace *testing.T with testing.TestingT (#908)

The WithDefaultRetryableErrors function now takes a testing.TestingT
argument instead of *testing.T to allow for anything that implements the
testing.TestingT interface to be used. This allows for GinkgoT() to be
used in tests utilizing the ginkgo testing framework.

falmotlag committed:
add k8s.GetKubernetesClusterVersion to get cluster version (#876)

* add k8s.GetKubernetesClusterVersion to get cluster version

* rename function

Co-authored-by: Yoriyasu Yano <430092+yorinasub17@users.noreply.github.com>

* Add a version function that takes options

* add unit tests

* test by comparing cluster version to the one returned by kubectl

Co-authored-by: Yoriyasu Yano <430092+yorinasub17@users.noreply.github.com>

davidtam committed:
Update terragrunt support to use terragrunt run-all plan / apply / destroy #896

davidtam committed:
adding run-all to TerraformCommandsWithLockSupport

davidtam committed:
set TF_LOG=DEBUG so that -lock CLI option passed down for assertion

davidtam committed:
adding back apply-all / etc to parallel / lock list for backward support

davidtam committed:
ran goimports

thundering-herd committed:
feat: add AWS SSO support (#883)

* feat: add AWS SSO support

To support AWS SSO we need to Bump the AWS SDK which is providing
support recently.

Also we need to adjust our creation of sessins to include the
`SharedConfigState` field. See
https://aws.amazon.com/blogs/developer/aws-sso-support-in-the-aws-sdk-for-go/
for more information.

* fix: gofmt

* Limit parallelism due to aws-sdk-go failing

Co-authored-by: Yoriyasu Yano <430092+yorinasub17@users.noreply.github.com>

GennadySpb committed:
Load and Save SSH KeyPair (#724)

dependabot[bot] committed:
Bump nokogiri from 1.11.1 to 1.11.5 in /docs (#907)

Bumps [nokogiri](https://github.com/sparklemotion/nokogiri) from 1.11.1 to 1.11.5.
- [Release notes](https://github.com/sparklemotion/nokogiri/releases)
- [Changelog](https://github.com/sparklemotion/nokogiri/blob/main/CHANGELOG.md)
- [Commits](https://github.com/sparklemotion/nokogiri/compare/v1.11.1...v1.11.5)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

cbuto committed:
Add new function to allow explicitly using in-cluster auth (#901)

Add comment

Update log messages

HadwaAbdelhalem committed:
expand azure resource group module (#912)

* expand azure resource group module

* fix resourcegroup_test

* fix resourcegroup.go test

HadwaAbdelhalem committed:
Update az keyvault cert (#921)

* update certificate file

* delete expired cert

HadwaAbdelhalem commented on modules/azure/automationaccount.go:
```diff
@@ -0,0 +1,342 @@
+
+// AutomationAccountDscCompiled indicates whether the specified Azure Automation Account DSC compiled successfully.
+// This function would fail the test if there is an error.
+func AutomationAccountDscCompiled(t testing.TestingT, dscConfiguraitonName string, automationAccountName string, resourceGroupName string, subscriptionID string) bool {
```
function name is a bit confusing, please rename it to AutomationAccountIsDscCompiled

robcamer replied:
Fixed in latest commit.

HadwaAbdelhalem commented on modules/azure/automationaccount.go:
```diff
@@ -0,0 +1,342 @@
+// AutomationAccountDscCompiledE indicates whether the specified Azure Automation Account DSC compiled successfully.
+// DSC compilation is performed via a Terraform null_resource using PowerShell Core, executing async.
+// Compilation can take a few minutes to spin up resources requiring a retry mechanism to allow for this
+func AutomationAccountDscCompiledE(t testing.TestingT, dscConfiguraitonName string, automationAccountName string, resourceGroupName string, subscriptionID string) (bool, error) {
```
function name is a bit confusing, please rename it to AutomationAccountIsDscCompiledE

robcamer replied:
Fixed in latest commits.

HadwaAbdelhalem commented on modules/azure/automationaccount.go:
```diff
@@ -0,0 +1,342 @@
+		}
+		sleep = 2 * sleep
+	}
+	return false, nil
```
add code comment under what condition would false be returned

robcamer replied:
Fixed in latest commit.

HadwaAbdelhalem commented on modules/azure/automationaccount.go:
```diff
@@ -0,0 +1,342 @@
+
+// AutomationAccountRunAsConnectionValidates indicates whether the specified Azure Automation Account RunAs Account exists.
+// This function would fail the test if there is an error.
+func AutomationAccountRunAsConnectionValidates(t testing.TestingT, automationAccountrunAsAccountName string, runAsConnectionType string, runAsCertificateThumbprint string, automationAccountName string, resourceGroupName string, subscriptionID string) bool {
```
name is confusing!, is there a more accurate name we can use, may be AutomationAccountHasARunAsConnection

robcamer replied:
Fixed in latest commit.

HadwaAbdelhalem commented on modules/azure/automationaccount.go:
```diff
@@ -0,0 +1,342 @@
+/////////
+
+// GetAutomationAccountE returns the Azure Automation Account by name if it exists in the subscription
+func GetAutomationAccountE(t testing.TestingT, automationAccountName string, resourceGroupName string, subscriptionID string) (*automation.Account, error) {
```
I would add a nonE version to those methods to open it up for general use

robcamer replied:
Added nonE versions of the methods for general use.

HadwaAbdelhalem reviewed (changes requested):
@robcamer Thanks Rob, few comments to resolve and would be ready to grunt team review

robcamer committed:
Updates based on PR review.  Test successfully locally.

robcamer commented on examples/azure/terraform-azure-automationaccount-example/outputs.tf:
```diff
@@ -0,0 +1,9 @@
+}
+output "sample_dsc_name" {
+  value = azurerm_automation_dsc_configuration.SampleDSC.name
+}
```
Fixed in latest commit.

robcamer commented on examples/azure/terraform-azure-automationaccount-example/README.md:
```diff
@@ -0,0 +1,66 @@
+# Terraform Azure Automation Account Example
+
+This folder contains a simple Terraform module that deploys resources in [Azure](https://azure.microsoft.com/) to demonstrate how you can use Terratest to write automated tests for your Azure Terraform code. This module deploys the following:
```
Fixed in latest commit.

robcamer commented on examples/azure/terraform-azure-automationaccount-example/README.md:
```diff
@@ -0,0 +1,66 @@
+  - `Automation Account Connection Run As Account` with the name specified in the `automation_run_as_connection_name` variable.  See the section titled *_Example Service Principal and Certificate Setup_* below on how to configure the Automation Account RunAs Account credentials.
+  - `Automation Account Connection Run As Certificate` with the name specified in the `automation_run_as_certificate_name` variable and the thumbprint in the `TF_VAR_automation_run_as_certificate_thumbprint` variable.
+  - `Automation Account Connection Type` with type specified in the `automation_run_as_connection_type` variable.
+  - [Desired State Configuration](https://docs.microsoft.com/en-us/powershell/scripting/dsc/getting-started/winGettingStarted?view=powershell-7#:~:text=Get%20started%20with%20Desired%20State%20Configuration%20%28DSC%29%20for,Windows%20PowerShell%20Desired%20State%20Configuration%20log%20files.%20) with the name specified in the `sample_dsc_name` variable and the path to the DSC specified in the ` ...
```
Fixed in latest commit.

robcamer committed:
added nonE methods for helpers

hattan commented:
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/aztfmod/terratest/actions/runs/934339910

hattan commented:
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/aztfmod/terratest/actions/runs/934393839

robcamer commented:
CI completed successfully after two additional commits to resolve feedback: https://github.com/aztfmod/terratest/actions/runs/934393839  (running the pipelines in this branch, pull-request-658)

HadwaAbdelhalem reviewed (dismissed):
Thank you @robcamer , @yorinasub17  ready for your review.

yorinasub17 commented on examples/azure/terraform-azure-automationaccount-example/main.tf:
```diff
@@ -0,0 +1,197 @@
+# # Compilation is triggered on every one to ensure the latest changes are always applied
+# # WARNING: ConvertTo-SecureString -String '${var.POWERSHELL_CLIENT_SECRET}, will expose the actual secret value in the logs.
+# # ---------------------------------------------------------------------------------------------------------------------
+resource "null_resource" "azureSignInPWSH" {
```
Resource names should be underscore based (e.g., `azure_signing_pwsh`) for consistency with other examples. This applies to all resources in this example.

yorinasub17 commented on examples/azure/terraform-azure-automationaccount-example/variables.tf:
```diff
@@ -0,0 +1,125 @@
+
+variable "ARM_SUBSCRIPTION_ID" {}
+variable "ARM_TENANT_ID" {}
+variable "client_id" { sensitive = true }
+variable "client_secret" { sensitive = true }
+variable "automation_run_as_certificate_thumbprint" {}
+variable "automation_run_as_certificate_base64" {}
```
All variables should have a description and type specified

yorinasub17 commented on examples/azure/terraform-azure-automationaccount-example/variables.tf:
```diff
@@ -0,0 +1,125 @@
+# You must provide a value for each of these parameters.
+# ---------------------------------------------------------------------------------------------------------------------
+
+variable "ARM_SUBSCRIPTION_ID" {}
+variable "ARM_TENANT_ID" {}
```
I don't think these two variables need to be terraform input vars, right? Since they aren't referenced directly in the module?

yorinasub17 commented on modules/azure/automationaccount.go:
```diff
@@ -0,0 +1,342 @@
+	return compiled
+}
+
+// AutomationAccountIsDscCompiledE indicates whether the specified Azure Automation Account DSC compiled successfully.
+// DSC compilation is performed via a Terraform null_resource using PowerShell Core, executing async.
+// Compilation can take a few minutes to spin up resources requiring a retry mechanism to allow for this
+func AutomationAccountIsDscCompiledE(t testing.TestingT, dscConfiguraitonName string, automationAccountName string, resourceGroupName string, subscriptionID string) (bool, error) {
+	seconds := 30 // 30 second initial delay
+	sleep := time.Duration(seconds) * time.Second
+	maxAttempts := 5 // try 5 times, doubling the delay each time.
+
+	for i := 1; i < maxAttempts; i++ {
+		time.Sleep(sleep)
+		dscCompileJobStatus, err := AutomationAccountDscCompileJobStatusE(t, dscConfiguraitonName, automationAccountName, resourceGroupName, subscriptionID)
+		if err != nil {
+			return false, err
+		}
+		// check if status === Completed
+		if dscCompileJobStatus == "Completed" {
+			return true, nil
+		}
+		sleep = 2 * sleep
+	}
+	return false, nil // Return false if the DSC compilation job completed but failed to compile such as from an error in the DSC code
+}
```
This should either be renamed to `WaitUntilDscCompiled`, or the retry logic should be removed. I think it is unintuitive for an `IsXYZ` function to be waiting on some condition.

Also, we already have a retry helper ([retry.DoWithRetry](https://github.com/gruntwork-io/terratest/blob/master/modules/retry/retry.go#L53)), which should be reused here.

yorinasub17 commented on modules/azure/automationaccount.go:
```diff
@@ -0,0 +1,342 @@
+	return *certificate.CertificateProperties.Thumbprint == runAsCertificateThumbprint, nil
+}
+
+// AutomationAccountHasARunAsConnection indicates whether the specified Azure Automation Account RunAs Account exists.
+// This function would fail the test if there is an error.
+func AutomationAccountHasARunAsConnection(t testing.TestingT, automationAccountrunAsAccountName string, runAsConnectionType string, runAsCertificateThumbprint string, automationAccountName string, resourceGroupName string, subscriptionID string) bool {
+	exists, err := AutomationAccountHasARunAsConnectionE(t, automationAccountrunAsAccountName, runAsConnectionType, runAsCertificateThumbprint, automationAccountName, resourceGroupName, subscriptionID)
+	require.NoError(t, err)
+
+	return exists
+}
+
+// AutomationAccountHasARunAsConnectionE indicates whether the specified Azure Automation Account RunAs Account exists.
+func AutomationAccountHasARunAsConnectionE(t testing.TestingT, automationAccountrunAsAccountName string, runAsConnectionType string, runAsCertificateThumbprint string, automationAccountName string, resourceGroupName string, subscriptionID string) (bool, error) {
+	runAsAccountConnection, err := GetAutomationAccountRunAsConnectionE(t, automationAccountrunAsAccountName, automationAccountName, resourceGroupName, subscriptionID)
+	if err != nil {
+		if ResourceNotFoundErrorExists(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return *runAsAccountConnection.Name == automationAccountrunAsAccountName &&
+		*runAsAccountConnection.ConnectionProperties.ConnectionType.Name == runAsConnectionType &&
+		*runAsAccountConnection.ConnectionProperties.FieldDefinitionValues["CertificateThumbprint"] == runAsCertificateThumbprint, nil
+}
```
Let's be consistent in function naming:

```suggestion
// AutomationAccountRunAsConnectionExists indicates whether the specified Azure Automation Account RunAs Account exists.
// This function would fail the test if there is an error.
func AutomationAccountRunAsConnectionExists(t testing.TestingT, automationAccountrunAsAccountName string, runAsConnectionType string, runAsCertificateThumbprint string, automationAccountName string, resourceGroupName string, subscriptionID string) bool {
	exists, err := AutomationAccountRunAsConnectionExistsE(t, automationAccountrunAsAccountName, runAsConnectionType, runAsCertificateThumbprint, automationAccountName, resourceGroupName, subscriptionID)
	require.NoError(t, err)

	return exists
}

// AutomationAccountRunAsConnectionExistsE indicates whether the specified Azure Automation Account RunAs Account exists.
func AutomationAccountRunAsConnectionExistsE(t testing.TestingT, automationAccountrunAsAccountName string, runAsConnectionType string, runAsCertificateThumbprint string, automationAccountName string, resourceGroupName string, subscriptionID string) (bool, error) {
	runAsAccountConnection, err := GetAutomationAccountRunAsConnectionE(t, automationAccountrunAsAccountName, automationAccountName, resourceGroupName, subscriptionID)
	if err != nil {
		if ResourceNotFoundErrorExists(err) {
			return false, nil
		}
		return false, err
	}
	return *runAsAccountConnection.Name == automationAccountrunAsAccountName &&
		*runAsAccountConnection.ConnectionProperties.ConnectionType.Name == runAsConnectionType &&
		*runAsAccountConnection.ConnectionProperties.FieldDefinitionValues["CertificateThumbprint"] == runAsCertificateThumbprint, nil
}
```

yorinasub17 reviewed (commented):
Mostly LGTM! I did have a few style comments though.

Commit:
Update modules/azure/automationaccount.go

Co-authored-by: Yoriyasu Yano <430092+yorinasub17@users.noreply.github.com>

Commit:
updates based on PR feedback.

Commit:
Merge branch 'master' into pull-request-658

Commit:
fix workflow flagged issues

robcamer commented:
@HadwaAbdelhalem I ran the CI.  There are errors for AKS and AvailabilitySet, but not Automation Account:  https://github.com/aztfmod/terratest/runs/4051679037?check_suite_focus=true.  How should we proceed?

yorinasub17 reviewed (dismissed):
Updates LGTM! Will wait on @HadwaAbdelhalem to approve the test results prior to merging this in.

yorinasub17 commented on modules/azure/resourcegroup.go:
```diff
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"testing"
 
+
```
Ah this extra whitespace is causing our go format check to fail. Can you remove this?

HadwaAbdelhalem reviewed (dismissed):
LGTM, thanks @robcamer I checked and the AKS failure is not a blocker, @yorinasub17  we are good to merge.

Commit:
Merge branch 'master' into pull-request-658

robcamer commented:
@HadwaAbdelhalem @yorinasub17.  Resolved the merge conflict.  I think this is ready.  Thanks!

robcamer commented:
@HadwaAbdelhalem @yorinasub17. Just checking if anything else is required?  Thanks!

denis256 commented:
Hi,
looks like `goimports` should be applied

```
goimports................................................................Failed
- hook id: goimports
- files were modified by this hook

test/azure/terraform_azure_automationaccount_example_test.go
test/azure/terraform_azure_automationaccount_example_test.go
modules/azure/automationaccount_test.go
modules/azure/resourcegroup.go
```

james00012 commented:
Thanks for the contribution, and apologies for the long delay. This PR has been open for over a year without movement, so we're closing it as stale. Please feel free to reopen if you find time to pick it back up again. We appreciate the effort.
